### PR TITLE
Added Find and FindStrict methods to generic lists

### PR DESCRIPTION
**Pull request by jocgir**
Moved BaseHelper methods in the right file
Added strict mode on list search functions/methods
Added Find and FindStrict methods to generic lists
Added RemoveNil and RemoveEmpty methods to generic lists
Added test to cover new functions
Simplified the test structure

### CODE REVIEW

**Comment by jocgir** [comment]
This look a big change, but in fact, a lot of this code is automatically generated by gogenerate.

File: collections/implementation/generic_test.go [hunk 1]
```diff
@@ -228,28 +203,25 @@ func Test_CreateList(t *testing.T) {
...
+		wantErr string
 	}{
-		{"Empty", nil, baseList{}, false},
-		{"With nil elements", []int{10}, make(baseList, 10), false},
-		{"With capacity", []int{0, 10}, make(baseList, 0, 10), false},
-		{"Too much args", []int{0, 10, 1}, nil, true},
+		{"Empty", nil, baseList{}, ""},
+		{"With nil elements", []int{10}, make(baseList, 10), ""},
+		{"With capacity", []int{0, 10}, make(baseList, 0, 10), ""},
+		{"Too much args", []int{0, 10, 1}, nil, "CreateList only accept 2 arguments, size and capacity"},
```
Comment (dblanchette) [inline collections/implementation/generic_test.go:211]:
```suggestion
		{"Too many args", []int{0, 10, 1}, nil, "CreateList only accept 2 arguments, size and capacity"},
```

File: template/extra_data.go [hunk 2]
```diff
@@ -189,6 +203,8 @@ var dataFuncsHelp = descriptions{
 	"prepend":        "Push elements onto the front of a list, creating a new list.",
 	"rest":           "Gets the tail of the list (everything but the first item)",
 	"reverse":        "Produces a new list with the reversed elements of the given list.",
+	"removeEmpty":    "Returns a list will all empty element removed.",
```
Comment (ycanty) [inline template/extra_data.go]:
```suggestion
	"removeEmpty":    "Returns a list with all empty elements removed.",
```

File: collections/implementation/generic_test.go [hunk 3]
```diff
@@ -314,52 +280,70 @@ func Test_list_CreateDict(t *testing.T) {
...
 		want    baseIDict
-		wantErr bool
+		wantErr string
 	}{
-		{"Empty", nil, nil, baseDict{}, false},
-		{"With capacity", nil, []int{10}, baseDict{}, false},
-		{"With too much parameter", nil, []int{10, 1}, nil, true},
+		{"Empty", nil, nil, baseDict{}, ""},
+		{"With capacity", nil, []int{10}, baseDict{}, ""},
+		{"With too much parameter", nil, []int{10, 1}, nil, "CreateList only accept 1 argument for size"},
```
Comment (dblanchette) [inline collections/implementation/generic_test.go:287]:
```suggestion
		{"With too many parameters", nil, []int{10, 1}, nil, "CreateList only accept 1 argument for size"},
```

File: template/extra_data.go [hunk 4]
```diff
@@ -189,6 +203,8 @@ var dataFuncsHelp = descriptions{
...
+	"removeNil":      "Returns a list will all nil element removed.",
```
Comment (ycanty) [inline template/extra_data.go]:
```suggestion
	"removeNil":      "Returns a list with all nil elements removed.",
```

File: template/extra_data.go [hunk 5]
```diff
@@ -163,12 +174,15 @@ var dataFuncsHelp = descriptions{
 	"array":          "Ensures that the supplied argument is an array (if it is already an array/slice, there is no change, if not, the argument is replaced by []interface{} with a single value).",
 	"bool":           "Converts the `string` into boolean value (`string` must be `True`, `true`, `TRUE`, `1` or `False`, `false`, `FALSE`, `0`)",
 	"char":           "Returns the character corresponging to the supplied integer value",
-	"contains":       "Test to see if a list has a particular elements.",
+	"contains":       "Test to see if a list has a particular elements (matches any types).",
```
Comment (ycanty) [inline template/extra_data.go:177]:
```suggestion
	"contains":       "Tests whether a list contains all given elements (matches any types).",
```

File: template/extra_data.go [hunk 6]
```diff
@@ -163,12 +174,15 @@ var dataFuncsHelp = descriptions{
...
+	"containsStrict": "Test to see if a list has a particular elements (matches only the same types).",
```
Comment (ycanty) [inline template/extra_data.go:178]:
```suggestion
	"containsStrict": "Tests whether a list contains all given elements (matches only the same types).",
```

File: template/extra_data.go [hunk 7]
```diff
@@ -163,12 +174,15 @@ var dataFuncsHelp = descriptions{
...
 	"content":        "Returns the content of a single element map.\nUsed to retrieve content in a declaration like:\n    value \"name\" { a = 1 b = 3 }",
 	"data":           "Tries to convert the supplied data string into data structure (Go spec). It will try to convert HCL, YAML and JSON format. If context is omitted, default context is used.",
 	"dict":           "Returns a new dictionary from a list of pairs (key, value).",
-	"extract":        "Extracts values from a slice or a map, indexes could be either integers for slice or strings for maps",
-	"get":            "Returns the value associated with the supplied map, key and map could be inverted for convenience (i.e. when using piping mode)",
+	"extract":        "Extracts values from a slice or a map, indexes could be either integers for slice or strings for maps.",
+	"find":           "Returns all index position where the element is found in the list (matches any types).",
```
Comment (ycanty) [inline template/extra_data.go:183]:
```suggestion
	"find":           "Returns all index positions where the element is found in the list (matches any types).",
```

File: template/extra_data.go [hunk 8]
```diff
@@ -163,12 +174,15 @@ var dataFuncsHelp = descriptions{
...
+	"findStrict":     "Returns all index position where the element is found in the list (matches only the same types).",
```
Comment (ycanty) [inline template/extra_data.go:184]:
```suggestion
	"findStrict":     "Returns all index positions where the element is found in the list (matches only the same types).",
```

File: template/extra_data.go (hunk 2 above)
Comment (dblanchette) [inline template/extra_data.go:206]:
```suggestion
	"removeEmpty":    "Returns a list will all empty elements removed.",
```

File: template/extra_data.go (hunk 4 above)
Comment (dblanchette) [inline template/extra_data.go:207]:
```suggestion
	"removeNil":      "Returns a list will all nil elements removed.",
```

**Pull request merged**